File: backend/app/core/tiers.py
"""Single source of truth for Binance P2P merchant tiers.

Every place that DETECTS or STORES a merchant tier goes through here, so the tier
set — and the special rule that a Block merchant keeps Gold capabilities — live in
ONE place. Change it here and every path (tier poller, connect-Binance, detect-name)
stays in sync.
"""

# Tiers we detect from Binance's ad feed, best -> worst. 'block' (the purple
# diamond — "Block Merchant") is Binance's TOP P2P tier, above gold.
DETECTABLE_TIERS = ("block", "gold", "silver", "bronze")


def merchant_tier_for(p2p_tier):
    """The binance_merchant_tier value to STORE for a detected p2p tier.

    Block merchants are ABOVE gold, but every Gold-gated capability
    (counterparty filters, the pro_max plan auto-assign, the 0.25 KES fee) checks
    `binance_merchant_tier == 'gold'`. So we store 'gold' for a block merchant —
    they keep all the Gold powers — while `binance_p2p_tier` records the true
    'block' medal, which is what the badge reads.

    Returns None for 'normal'/unknown, meaning "don't touch merchant_tier".
    """
    if p2p_tier == "block":
        return "gold"
    if p2p_tier in ("gold", "silver", "bronze"):
        return p2p_tier
    return None


def display_tier(trader):
    """A trader's DISPLAY tier (block/gold/silver/bronze/None) — 'block' from the
    real P2P medal, else the capability tier. This is what prices the weekly plan
    and drives the badge. Mirrors frontend merchantDisplayTier()."""
    p2p = (getattr(trader, "binance_p2p_tier", None) or "").lower()
    if p2p == "block":
        return "block"
    mt = (getattr(trader, "binance_merchant_tier", None) or getattr(trader, "binance_p2p_tier", None) or "").lower()
    return mt or None

File: frontend/src/config/tiers.js
// Single source of truth for merchant-tier colours, labels and the ◆ Block badge
// across the WHOLE app — Market Activity, Price Tracker, the Settings badge and
// the Admin traders list. Change a colour / add a tier here and every surface
// updates. (Backend counterpart: app/core/tiers.py.)

export const TIER_COLOR = {
  block:  '#A855F7',   // purple — Binance Block Merchant (top tier)
  gold:   '#FFBE52',
  silver: '#D6DBE2',
  bronze: '#F08A3C',
  normal: '#929AA6',
};

export const TIER_LABEL = {
  block: 'Block', gold: 'Gold', silver: 'Silver', bronze: 'Bronze', normal: 'Normal',
};

// The ◆ diamond marks the Block tier everywhere it appears.
export const TIER_DIAMOND = '◆';

export const tierColor = (t) => TIER_COLOR[String(t || '').toLowerCase()] || TIER_COLOR.normal;
export const tierLabel = (t) => TIER_LABEL[String(t || '').toLowerCase()] || (t || '');

// A merchant's DISPLAY tier from their stored fields. `binance_p2p_tier` is the
// true medal (incl. 'block'); `binance_merchant_tier` is the capability tier
// (block maps to 'gold'). Block wins so the badge always shows the upgrade.
export function merchantDisplayTier(m) {
  const p2p = String(m?.binance_p2p_tier || '').toLowerCase();
  if (p2p === 'block') return 'block';
  return String(m?.binance_merchant_tier || m?.binance_p2p_tier || '').toLowerCase();
}
